pkg/util/ioutil/ioutil.go

```diff
@@ -7,69 +7,70 @@ package ioutil
 import (
 	"context"
 	"io"
+	"sync/atomic"
 	"time"
 )
 
 // timeoutReader monitors the progress of the io.Reader Read method
-// and calls the cancel function when reading does not progress for
+// and calls the cancelFn function when reading does not progress for
 // the specified time.
 type timeoutReader struct {
-	t *time.Timer
-	r io.Reader
-	n chan int
+	total  atomic.Uint64 // Total number of bytes read until timeout.
+	reader io.Reader
+
+	timer   *time.Timer
+	timeout time.Duration
+
+	done   atomic.Bool
+	cancel context.CancelFunc
 }
 
 // Read implements the io.Reader Read interface.
 func (tr *timeoutReader) Read(p []byte) (int, error) {
-	n, err := tr.r.Read(p)
-
-	v := n
-	if err != nil && tr.t.Stop() {
-		select {
-		case <-tr.t.C:
-		default:
-		}
-
-		// Negative value signals that error has
-		// occurred and the goroutine should terminate.
-		v = -1
-	}
-
-	select {
-	case tr.n <- v:
-	default:
+	n, err := tr.reader.Read(p)
+	switch {
+	case err != nil && !tr.done.Load():
+		tr.cancel()
+		tr.done.Store(true)
+	case !tr.done.Load():
+		tr.total.Add(uint64(n))
+		tr.timer.Reset(tr.timeout)
 	}
-
 	return n, err
```

Comment:
would it make sense to return non nil error when reader has timed out?

Reply:
this case would be following:

1. read some data (but there are more data to be read)
2. reader is now blocked
3. reader timeouts
4 if at this point reading from reader is attempted should it return non nil err without actually attempting to read from underlying reader?

Reply:
Fair enough.